index.js

```diff
@@ -1,125 +1,211 @@
 'use strict'
 
-const nodePath = require('path')
-
-const commonExtname = require('common-extname')
-const commonPathPrefix = require('common-path-prefix')
+const path = require('path')
+const escapeStringRegexp = require('escape-string-regexp')
+const GLOBSTAR = require('minimatch').GLOBSTAR
 const glob = require('glob')
 const identifierfy = require('identifierfy')
-const flatten = require('lodash.flatten')
 
-function getPattern (path, source) {
-  const pattern = source.value.replace(/^glob:/, '').trim()
-  if (!pattern) {
-    throw path.buildCodeFrameError(`Missing glob pattern '${source.value}'`)
-  }
-  if (pattern.startsWith('/')) {
-    throw path.buildCodeFrameError(`Glob pattern must be relative, was '${pattern}'`)
-  }
-  return pattern
+const twoStar = '(?:(?!(?:/|^)\\.).)*?' // match "**"
+
+function makeSubpathExpressions (minimatch) {
+  return minimatch.set.map((expressions, index) => {
+    const nextExpressions = minimatch.set[(index + 1) % minimatch.set.length]
+
+    const parts = []
+    let lastCaptureIndex = -1
+    for (let expressionIndex = 0; expressionIndex < expressions.length; expressionIndex++) {
+      const expression = expressions[expressionIndex]
+
+      let capture = true
+      let partialPattern
+      if (typeof expression === 'string') {
+        partialPattern = escapeStringRegexp(expression)
+        // `capture` should only be true if brace expansion is used, and … some
+        // other condition?
+        // FIXME: What is the logic here? {a,b} will lead to capture, as will
+        // {a,a*}. Is this to do with multiple braces being expanded? E.g.
+        // {a,b}/foo/{c,d}?
+        capture = expression !== nextExpressions[expressionIndex]
+      } else if (expression === GLOBSTAR) {
+        partialPattern = twoStar
+      } else {
+        partialPattern = expression.source.slice(1, -1)
+      }
+
+      if (capture) {
+        if (lastCaptureIndex < 0) {
+          partialPattern = '(' + partialPattern
+        }
+        lastCaptureIndex = expressionIndex
+      }
+      parts.push(partialPattern)
+    }
+    parts[lastCaptureIndex] += ')'
+
+    const last = expressions.length - 1
+    // FIXME: Why check for strings? What is the significance of the expressions[last] !== nextExpressions[last]?
+    // Why not use path.extname()?
+    const extname = typeof expressions[last] !== 'string' || expressions[last] !== nextExpressions[last]
+      // grab extension from filename
+      ? minimatch.globParts[index][last].match(/(?:\.[A-Za-z0-9]+)*$/)[0]
+      : null
+
+    return {
+      // FIXME: Why even return extname? Can't it be excluded from the regexp match?
```

Comment:
Hmmmm.... I mean we could by using `minimatch.set` instead. I'm just concerned the regexp could change. `minimatch.globParts` is just the brace-expanded glob pattern, which should stay consistent. If we were to use `minimatch.set`, we would have to de-duplicate the expressions.